pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+        type CurrentSessionIndex: GetSessionIndex<SessionIndex>;
+
+        /// Helper that returns the block author for the orchestrator chain (if it exists)
+        type GetSelfChainBlockAuthor: MaybeSelfChainBlockAuthor<Self::CollatorId>;
+    }
+
+    /// Switch to enable/disable inactivity tracking
+    #[pallet::storage]
+    pub type EnableInactivityTracking<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    /// A list of double map of inactive collators for a session
+    #[pallet::storage]
+    pub type ActiveCollators<T: Config> = StorageMap<
+        _,
+        Twox64Concat,
+        SessionIndex,
+        BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>,
+        ValueQuery,
+    >;
+
+    /// A list of inactive collators for a session. Repopulated at the start of every session
+    #[pallet::storage]
+    pub type ActiveCollatorsForCurrentSession<T: Config> =
+        StorageValue<_, BoundedVec<T::CollatorId, T::MaxCollatorsPerSession>, ValueQuery>;
+
+    /// The last session index for which the inactive collators have not been processed
+    #[pallet::storage]
+    pub type LastUnprocessedSession<T: Config> = StorageValue<_, SessionIndex, ValueQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        InactivityTrackingEnabled { is_enabled: bool },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        MaxCollatorsPerSessionReached,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::DbWeight::get().reads_writes(0, 1))]
+        pub fn set_inactivity_tracking_status(
+            origin: OriginFor<T>,
+            is_enabled: bool,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            <EnableInactivityTracking<T>>::put(is_enabled);
+            Self::deposit_event(Event::<T>::InactivityTrackingEnabled { is_enabled });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists)
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight += Self::on_author_noted(orchestrator_chain_author);
+            }
+
+            // Update inactive collator records only after a session has ended
+            let current_session = T::CurrentSessionIndex::session_index();
+            let current_unprocessed_session = <LastUnprocessedSession<T>>::get();
+            if current_unprocessed_session < current_session {
+                total_weight +=
+                    Self::process_ended_session(current_unprocessed_session, current_session);
```

Comment:
We said this could be done through pallet-intiliazer right?

Reply:
and indeed this does not work in my opinion, you should see it in the tests. Here you will enter when the new session starts, which means that your `ActiveCollatorsForCurrentSession` will be of those when the session changes, and not before. you need to do this BEFORE the next session starts. so either you predict that is going to change or do it through pallet-initializer

Reply:
Yup plan to refactor this one

Reply:
Refactored the function in https://github.com/moondance-labs/tanssi/pull/922/commits/366077535ba21025f2eab141774aa71fcbd648b6 and added it to `apply_new_session` in https://github.com/moondance-labs/tanssi/pull/922/commits/213e28b88a5dccb8988077aa160c6f0dc57fb578